substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Target_java_lang_ClassLoader.java

```diff
@@ -215,6 +218,33 @@ private URL getResource(String name) {
     @Alias //
     private static ClassLoader scl;
 
+    @Inject
+    @RecomputeFieldValue(kind = Kind.Custom, declClass = ClassLoaderJfrIDRecomputation.class)
+    public int jfrID;
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    static HashMap<ClassLoader, Integer> jfrIdsMap = new HashMap<>();
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public static void setJfrID(ClassLoader cl, Integer id) {
+        jfrIdsMap.put(cl, id);
+    }
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    static class ClassLoaderJfrIDRecomputation implements RecomputeFieldValue.CustomFieldValueComputer {
+        @Override
+        public Object compute(MetaAccessProvider metaAccess, ResolvedJavaField original, ResolvedJavaField annotated, Object receiver) {
+            ClassLoader cl = (ClassLoader) receiver;
+            Integer val = jfrIdsMap.get(cl);
+            if (val == null) {
+                // Why are some items not registered?
```

Comment:
Hmm ok, but that means we have to reserve the 0 slot in the trace-ID table for possible use by those objects? My thinking was that because it's not relevant for JFR anyway it may be safer to fail. Which would indicate that it *is* relevant after all, and we need to do something about it, rather than let it silently pick up the trace-ID of the wrong type. Could we let new objects (e.g. classloaders) at runtime let their jfrId fields initialize with -1?

Reply:
With "not relevant", I meant that such an object should never be visible to JFR. It would be good to assert/guarantee that in some way as picking up the wrong trace id silently is rather dangerous.

When JFR is disabled, we want the static analysis to eliminate the extra `jfrId` fields. At the moment, this should work just fine as the static analysis does not encounter any runtime usages of the `jfrId` fields (at runtime, the field is only accessed in JFR code, which is unreachable when `AllowVMInspection` is disabled).

So, using 0 as uninitialized would be the easiest way to achieve that. Using -1 as uninitialized might work as well but it at least requires substituting all the constructors. This is JDK-specific and therefore it is easy to miss a constructor.